### PR TITLE
Feature/add custom matchers

### PR DESCRIPTION
**Pull request by barakyoresh**

### CODE REVIEW

File: LLReactiveMatchers/Helpers/LLSignalTestRecorder.m [hunk 1]
```diff
@@ -33,33 +33,34 @@ @interface LLSignalTestRecorder ()
...
 
 - (id) init {
-    if( (self = [super init]) ) {
-        self.receivedEvents = [NSMutableArray array];
-        self.activeThreadsInReceivedEvents = [NSMutableSet set];
-        self.passthrough = [RACReplaySubject replaySubjectWithCapacity:RACReplaySubjectUnlimitedCapacity];
-        [self startCountingSubscriptions];
-    }
-    return self;
+  if( (self = [super init]) ) {
```
Comment (StatusReport) [inline LLReactiveMatchers/Helpers/LLSignalTestRecorder.m]: Please keep his original 4 spaces convention - there are too many redundant lines that were changed here.

File: LLReactiveMatchers/Matchers/EXPMatchers+deliverValuesOnThread.h [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+#import <Expecta/Expecta.h>
+
+/// Checks if actual \c LLSignalTestRecorder or \c RACSignal delivered values with \c expected as
+/// the current thread.
+/// Common usage:
+/// @code
+/// RACSignal *mySignal = [SignalReturningClass signalReturningMethod];
+/// NSThread *myThread = [SignalReturningClass expectedDeliveryThread];
+///
+/// expect(mySignal).will.deliverValuesThread(myThread);
```
Comment (StatusReport) [inline LLReactiveMatchers/Matchers/EXPMatchers+deliverValuesOnThread.h]: deliverValuesOnThread

File: LLReactiveMatchersTests/Tests/EXPMatchers+deliverValuesOnThread.m [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+  signal = [signal asyncySignal];
+  assertPass(test_expect(signal).willNot.deliverValuesOnThread(thread));
+  assertFail(test_expect(signal).will.deliverValuesOnThread(thread), failureString);
+
+  assertPass(test_expect(signal).willNotContinueTo.deliverValuesOnThread(thread));
+  assertFail(test_expect(signal).willContinueTo.deliverValuesOnThread(thread), failureString);
+}
+
+- (void) test_2ValuesCorrect {
```
Comment (StatusReport) [inline LLReactiveMatchersTests/Tests/EXPMatchers+deliverValuesOnThread.m]: I don't know if that conforms LL style, but I prefer `two` over `2`.

File: LLReactiveMatchersTests/Tests/EXPMatchers+deliverValuesOnThread.m [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+- (void) test_2ValuesinCorrect {
+  RACSubject *subject = [RACSubject subject];
+  NSThread *thread = [[NSThread alloc] init];
+  RACSignal *replayMultiThreadedSignal = [[[subject
+                                         deliverOn:[RACScheduler scheduler]]
+                                         merge:subject]
+                                         replay];
+  [subject sendNext:@"0"];
+  NSString *failureString = [NSString stringWithFormat:@"First failure: expected values to be "
+                             "delivered on thread %@ but (\n    0,\n    0\n) were delivered on "
```
Comment (StatusReport) [inline LLReactiveMatchersTests/Tests/EXPMatchers+deliverValuesOnThread.m]: `(\n    0,\n    0\n)` -> `%@` and the description of that object.

File: LLReactiveMatchersTests/Tests/EXPMatchers+sendErrorTest.m [hunk 5]
```diff
@@ -46,7 +46,7 @@ - (void) test_notYetCompleted {
 - (void) test_endsInSameError {
     RACSignal *signal = [[[LLReactiveMatchersFixtures values:@[@1, @2, @3]] concat:[RACSignal error:LLSpecError]] setNameWithFormat:@"foo"];
     NSError *error = LLSpecError;
-    NSString *failureString = @"expected: actual foo to send error Error Domain=com.github.lawrencelomax.llreactivematchers.fixture Code=0 \"The operation couldn’t be completed. (com.github.lawrencelomax.llreactivematchers.fixture error 0.)\", got: the same error";
+    NSString *failureString =  @"expected: actual foo to send error Error Domain=com.github.lawrencelomax.llreactivematchers.fixture Code=0 \"(null)\", got: the same error";
```
Comment (StatusReport) [inline LLReactiveMatchersTests/Tests/EXPMatchers+sendErrorTest.m]: Redundant ` `.

File: LLReactiveMatchersTests/Tests/EXPMatchers+sendErrorTest.m [hunk 6]
```diff
@@ -59,8 +59,8 @@ - (void) test_endsInSameError {
 - (void) test_endsInDifferentError {
     RACSignal *signal = [[[LLReactiveMatchersFixtures values:@[@1, @2, @3]] concat:[RACSignal error:LLSpecError]] setNameWithFormat:@"foo"];
     NSError *error = [NSError errorWithDomain:@"foo" code:1 userInfo:@{}];
-    NSString *failureString = @"expected: actual foo to send error Error Domain=foo Code=1 \"The operation couldn’t be completed. (foo error 1.)\", got: different errors";
-    
+    NSString *failureString = @"expected: actual foo to send error Error Domain=foo Code=1 \"(null)\", got: different errors";
```
Comment (StatusReport) [inline LLReactiveMatchersTests/Tests/EXPMatchers+sendErrorTest.m]: You can repeat the `%@` with `-[NSError description]` here as well.

**Comment by StatusReport** [comment]
Round 1.

File: LLReactiveMatchers/Helpers/LLSignalTestRecorder.m (hunk 1 above)
Comment (barakyoresh) [inline LLReactiveMatchers/Helpers/LLSignalTestRecorder.m]: Done, wasn't me actually, not sure what caused it.

File: LLReactiveMatchers/Matchers/EXPMatchers+deliverValuesOnThread.h (hunk 2 above)
Comment (barakyoresh) [inline LLReactiveMatchers/Matchers/EXPMatchers+deliverValuesOnThread.h]: Done.

File: LLReactiveMatchersTests/Tests/EXPMatchers+deliverValuesOnThread.m (hunk 3 above)
Comment (barakyoresh) [inline LLReactiveMatchersTests/Tests/EXPMatchers+deliverValuesOnThread.m]: It doesn't. But neither do lines of less than _two_ mile length. 
So done.

File: LLReactiveMatchersTests/Tests/EXPMatchers+deliverValuesOnThread.m (hunk 4 above)
Comment (barakyoresh) [inline LLReactiveMatchersTests/Tests/EXPMatchers+deliverValuesOnThread.m]: Done.

File: LLReactiveMatchersTests/Tests/EXPMatchers+sendErrorTest.m (hunk 5 above)
Comment (barakyoresh) [inline LLReactiveMatchersTests/Tests/EXPMatchers+sendErrorTest.m]: Done` `.

File: LLReactiveMatchersTests/Tests/EXPMatchers+sendErrorTest.m (hunk 6 above)
Comment (barakyoresh) [inline LLReactiveMatchersTests/Tests/EXPMatchers+sendErrorTest.m]: Done.

**Comment by barakyoresh** [comment]
Round 1 complete.

File: LLReactiveMatchersTests/Tests/EXPMatchers+deliverValuesOnThread.m [hunk 7]
```diff
@@ -0,0 +1,77 @@
+#import "EXPMatchers+deliverValuesOnThread.h"
+
+@interface EXPMatchers_deliverValuesOnThreadTests : TEST_SUPERCLASS
+@end
+
+@implementation EXPMatchers_deliverValuesOnThreadTests
+
+- (void) test_noValuesIncorrect {
+  RACSignal *signal = RACSignal.empty;
```
Comment (StatusReport) [inline LLReactiveMatchersTests/Tests/EXPMatchers+deliverValuesOnThread.m]: `[RACSignal empty]`

Comment (barakyoresh) [inline LLReactiveMatchersTests/Tests/EXPMatchers+deliverValuesOnThread.m]: Done.

**Comment by barakyoresh** [comment]
Fix and squashed.

File: LLReactiveMatchers/Helpers/LLSignalTestRecorder.m [hunk 8]
```diff
@@ -169,13 +170,13 @@ - (BOOL) haveFinished {
...
 - (NSSet *)operatingThreads {
-  @synchronized(self) {
-    return [self.activeThreadsInReceivedEvents copy];
-  }
+    @synchronized(self) {
+        return [self.activeThreadsInReceivedEvents copy];
+    }
 }
 
 - (NSUInteger) operatingThreadsCount {
```
Comment (amit-goldstein) [inline LLReactiveMatchers/Helpers/LLSignalTestRecorder.m]: any reason for the space after ) and before the method name on some of the methods in this file?

**Pull request merged**